File: emu/emu_plot/python/plot_atrb.py
# Read Attribution Tool output

import os
import numpy as np
import glob
import matplotlib.pyplot as plt
import global_emu_var as emu

def plot_atrb(frun): 
    # Define the byte order ('>' for big-endian)
    byte_order = '>'

    # Set EMU output file directory
    ff = os.path.join(frun, 'output')

    # specify controls in file 
    fctrl = ['lhs', 'wind', 'htflx', 'fwflx', 'sflx', 'pload', 'ic']
    nterms=len(fctrl)

    emu.atrb_ctrl=fctrl

    # ---------------
    # Read atrb.out_? (OBJF time-series)

    fdum = 'atrb.out_'
    fdum_all = fdum + '*'
    aa = glob.glob(os.path.join(ff, fdum_all))
    aa.sort()

    if len(aa) != 1:
        print()
        if len(aa) == 0:
            print('*********************************************')
            print(f'File {fdum_all} not found ... ')
            print('')
        else:
            print('*********************************************')
            print(f'More than one {fdum_all} found ... ')
            print('')
        return

    ip = aa[0].find(fdum) + len(fdum)
    frec = aa[0][ip:]
    nrec = int(frec)  # Length of time-series

    atrb = np.zeros((nterms, nrec), dtype=np.float32)
    atrb_mn = np.zeros(nterms, dtype=np.float32)

# Fortran/IDL are column-major, Python is row-major
    with open(aa[0], 'rb') as f:
        atrb = np.fromfile(f, dtype=byte_order+'f4', count=nrec * nterms).reshape((nterms, nrec))
        atrb_mn = np.fromfile(f, dtype=byte_order+'f4', count=nterms)
 
    emu.atrb = atrb
    emu.atrb_mn = atrb_mn

    print('*********************************************')
    print('Read OBJF and contributions to it from different controls')
    print('   atrb: temporal anomaly ')
    print('   atrb_mn: reference time-mean ')
    print('   fctrl: names of atrb/atrb_mn variables ')
    print(f'from file {aa[0]}')
    print()

    # ---------------
    # Read atrb.step_? (time-step)

    fdum = 'atrb.step_' + frec
    aa = glob.glob(os.path.join(ff, fdum))
    aa.sort()

    if len(aa) == 0:
        print('*********************************************')
        print(f'File {fdum} not found ... ')
        print('')
        return

    atrb_hr = np.zeros(nrec, dtype=np.int32)

    with open(aa[0], 'rb') as f:
        atrb_hr = np.fromfile(f, dtype=byte_order+'i4', count=nrec)

    emu.atrb_hr = atrb_hr

    print('*********************************************')
    print('Read variable ')
    print('   atrb_hr: sample time (hours from 1/1/1992 12Z)')
    print(f'from file {aa[0]}')
    print('')

    atrb_t = (atrb_hr / 24.0) / 365.0 + 1992.0

    # Plot
    frun_file = os.path.basename(frun)

    tmin = int(np.floor(np.min(atrb_t))) - 1
    tmax = int(np.ceil(np.max(atrb_t))) + 1

    plt.figure()
    plt.plot(atrb_t, atrb[0, :], label=fctrl[0])
    for i in range(1, nterms):
        plt.plot(atrb_t, atrb[i, :], label=fctrl[i])

    plt.title(frun_file)
    plt.xlabel('atrb_hr')
    plt.ylabel('atrb')
    plt.xlim(tmin, tmax)
    plt.autoscale(axis='y')
    plt.grid(True)
    plt.legend()

    plt.ion()  # Enable interactive mode
    plt.show(block=False)  # Show the plot without blocking

#    print("**************** HELLO WORLD  from atrb *******************")

File: emu/emu_plot/matlab/plot_atrb.m
function plot_atrb(frun)
    % PLOT_ATRB Read and plot Attribution Tool output from EMU
    %
    %   Input:
    %     frun - path to EMU run directory (e.g., 'emu_atrb_m_2_45_585_1')

    global emu
    byte_order = 'ieee-be';  % Big-endian
    ff = fullfile(frun, 'output');

    % ----------------------
    % List of control terms
    % ----------------------
    fctrl = {'ref', 'wind', 'htflx', 'fwflx', 'sflx', 'pload', 'ic', 'mean'};
    nterms = numel(fctrl);

    % ----------------------
    % Read atrb.out_? file
    % ----------------------
    files = dir(fullfile(ff, 'atrb.out_*'));
    if numel(files) ~= 1
        fprintf('\n*********************************************\n');
        if isempty(files)
            fprintf('File atrb.out_* not found ...\n\n');
        else
            fprintf('More than one atrb.out_* found ...\n\n');
        end
        return
    end

    fname = fullfile(ff, files(1).name);
    rec_tag = extractAfter(files(1).name, 'atrb.out_');
    nrec = str2double(rec_tag);

    % Check number of terms in file
    info = dir(fname);
    nbytes = info.bytes;
    nfloat = nbytes / 4;  % number of float32 values
    mterms = fix(nfloat / (nrec + 1));  % number of terms in file

    if mterms == 7
      fprintf('\n');
      fprintf('*********************************************\n');
      fprintf('!!!! BEWARE !!!!\n');
      fprintf('Chosen EMU output is that of an older version of Attribution Tool\n');
      fprintf('with 7 terms and excludes contributions from the mean.\n');
      fprintf('\n');
    end

    % Read binary data
    fid = fopen(fname, 'r', byte_order);
    atrb = fread(fid, [nrec, mterms], 'float32');
    atrb_mn = fread(fid, mterms, 'float32');
    fclose(fid);

    fprintf('*********************************************\n');
    fprintf('Read OBJF and contributions to it from different controls\n');
    fprintf('   atrb: temporal anomaly \n');
    fprintf('   atrb_mn: reference time-mean \n');
    fprintf('   atrb_ctrl: names of atrb/atrb_mn variables \n');
    fprintf('from file %s\n\n', fname);

    emu.atrb = atrb;
    emu.atrb_mn = atrb_mn;
    emu.atrb_ctrl = fctrl;

    % ----------------------
    % Read atrb.step_? file
    % ----------------------
    step_file = fullfile(ff, ['atrb.step_' rec_tag]);
    if ~isfile(step_file)
        fprintf('*********************************************\n');
        fprintf('File %s not found ...\n\n', ['atrb.step_' rec_tag]);
        return
    end

    fid = fopen(step_file, 'r', byte_order);
    atrb_hr = fread(fid, nrec, 'int32');
    fclose(fid);

    fprintf('*********************************************\n');
    fprintf('Read variable \n');
    fprintf('   atrb_hr: sample time (hours from 1/1/1992 12Z)\n');
    fprintf('from file %s\n\n', step_file);

    emu.atrb_hr = atrb_hr;

    % ----------------------
    % Time conversion and plot
    % ----------------------
    atrb_t = double(atrb_hr) / 24 / 365 + 1992;
    [~, frun_name] = fileparts(frun);

    tmin = floor(min(atrb_t)) - 1;
    tmax = ceil(max(atrb_t)) + 1;

    figure;
    hold on
    plot(atrb_t, atrb(:, 1), 'DisplayName', fctrl{1});
    for i = 2:mterms
	      plot(atrb_t, atrb(:, i), 'DisplayName', fctrl{i});
    end
    hold off
    legend show 
    title(frun_name, 'Interpreter', 'none');
    xlabel('Time (decimal year)');
    ylabel('atrb');
    xlim([tmin, tmax]);
    grid on;
end
